### ISSUE TITLE
False positive in `noUndeclaredVariables`

### ISSUE THREAD

**Issue by ah-yu** [biomejs/biome#2952]
### Environment information

```bash
CLI:
  Version:                      1.7.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v18.14.2"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 true

Linter:
  Recommended:                  false
  All:                          false
  Rules:                        a11y/noAccessKey = "error"
                                a11y/noAriaUnsupportedElements = "error"
                                a11y/noAutofocus = "error"
                                a11y/noBlankTarget = "error"
                                a11y/noDistractingElements = "error"
                                a11y/noHeaderScope = "error"
                                a11y/noInteractiveElementToNoninteractiveRole = "off"
                                a11y/noNoninteractiveElementToInteractiveRole = "warn"
                                a11y/noNoninteractiveTabindex = "off"
                                a11y/noPositiveTabindex = "error"
                                a11y/noRedundantAlt = "warn"
                                a11y/noRedundantRoles = "warn"
                                a11y/useAltText = "error"
                                a11y/useAnchorContent = "error"
                                a11y/useAriaActivedescendantWithTabindex = "error"
                                a11y/useAriaPropsForRole = "warn"
                                a11y/useHeadingContent = "error"
                                a11y/useHtmlLang = "error"
                                a11y/useIframeTitle = "error"
                                a11y/useKeyWithClickEvents = "warn"
                                a11y/useKeyWithMouseEvents = "warn"
                                a11y/useMediaCaption = "error"
                                a11y/useValidAnchor = "warn"
                                a11y/useValidAriaProps = "error"
                                a11y/useValidAriaRole = "warn"
                                a11y/useValidAriaValues = "off"
                                complexity/noExtraBooleanCast = "error"
                                complexity/noMultipleSpacesInRegularExpressionLiterals = "error"
                                complexity/noUselessCatch = "error"
                                complexity/noWith = "error"
                                correctness/noChildrenProp = "error"
                                correctness/noConstAssign = "error"
                                correctness/noConstantCondition = "error"
                                correctness/noEmptyCharacterClassInRegex = "error"
                                correctness/noEmptyPattern = "error"
                                correctness/noGlobalObjectCalls = "error"
                                correctness/noInnerDeclarations = "warn"
                                correctness/noInvalidConstructorSuper = "error"
                                correctness/noInvalidUseBeforeDeclaration = "error"
                                correctness/noNewSymbol = "error"
                                correctness/noNonoctalDecimalEscape = "error"
                                correctness/noPrecisionLoss = "error"
                                correctness/noSelfAssign = "error"
                                correctness/noSetterReturn = "error"
                                correctness/noSwitchDeclarations = "off"
                                correctness/noUndeclaredVariables = "error"
                                correctness/noUnreachable = "error"
                                correctness/noUnreachableSuper = "error"
                                correctness/noUnsafeFinally = "error"
                                correctness/noUnsafeOptionalChaining = "error"
                                correctness/noUnusedLabels = "error"
                                correctness/noUnusedVariables = "warn"
                                correctness/useExhaustiveDependencies = "warn"
                                correctness/useHookAtTopLevel = "error"
                                correctness/useIsNan = "error"
                                correctness/useJsxKeyInIterable = "error"
                                correctness/useValidForDirection = "error"
                                correctness/useYield = "error"
                                security/noDangerouslySetInnerHtml = "error"
                                security/noDangerouslySetInnerHtmlWithChildren = "error"
                                style/noImplicitBoolean = "off"
                                style/noVar = "warn"
                                style/useBlockStatements = "off"
                                style/useImportType = "error"
                                suspicious/noAssignInExpressions = "warn"
                                suspicious/noAsyncPromiseExecutor = "error"
                                suspicious/noCatchAssign = "error"
                                suspicious/noClassAssign = "error"
                                suspicious/noCommentText = "error"
                                suspicious/noCompareNegZero = "error"
                                suspicious/noControlCharactersInRegex = "error"
                                suspicious/noDebugger = "error"
                                suspicious/noDuplicateCase = "error"
                                suspicious/noDuplicateClassMembers = "error"
                                suspicious/noDuplicateJsxProps = "error"
                                suspicious/noDuplicateObjectKeys = "error"
                                suspicious/noDuplicateParameters = "error"
                                suspicious/noEmptyBlockStatements = "warn"
                                suspicious/noFallthroughSwitchClause = "warn"
                                suspicious/noFunctionAssign = "error"
                                suspicious/noGlobalAssign = "error"
                                suspicious/noImportAssign = "error"
                                suspicious/noMisleadingCharacterClass = "error"
                                suspicious/noPrototypeBuiltins = "warn"
                                suspicious/noRedeclare = "error"
                                suspicious/noShadowRestrictedNames = "off"
                                suspicious/noUnsafeNegation = "error"
                                suspicious/useGetterReturn = "error"
                                suspicious/useValidTypeof = "error"
```

### Rule name

noUndeclaredVariables

### Playground link

https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&quoteStyle=single&trailingComma=es5&semicolons=as-needed&attributePosition=multiline&bracketSameLine=true&lintRules=all&code=ZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABmAG8AbwAoACkAewAKACAAIABmAG8AbwAuAGEAcABwAGwAeQAoAHQAaABpAHMALAAgAGEAcgBnAHUAbQBlAG4AdABzACkACgB9AA%3D%3D

### Expected result

The `arguments` variable should not be reported as undeclared because the [arguments](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments) object is a local variable within non-arrow functions.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct
